### ISSUE TITLE
Connection object `closed` still zero after the [TCP/IP] connection is terminated

### ISSUE THREAD

**Issue by amn** [psycopg/psycopg2#311]
I have taken a look at the closely related issues [#263](https://github.com/psycopg/psycopg2/issues/263) and [#196](https://github.com/psycopg/psycopg2/issues/196).

Tested with psycopg2 2.6 as installed by `pip install psycopg2` in a virtual environment, on a CentOS 6.5, with Postgres 9.3 running on another host on the same network, also CentOS 6.5.

Doing:

```python
import psycopg2
conn = psycopg2.connect(...)
cur = conn.cursor()
cur.execute('select 1')
```

At which point I shutdown the remote Postgres server with a `service postgresql-9.3 stop`, and return to the above, typing:

```python
conn.closed
```

...which evaluates to `0`. I expect a closed connection and thus `1` instead. I also waited half an hour to see whether it was a lingering TCP/IP connection delaying necessary event(s) taking place, but it still got me `0` after all that time.

I can do additional testing, if needed.

I am going about this correctly, am I not? I just would like to somehow protect my client code from aborted connections, however these may manifest. Catching `OperationalError` exceptions is something I am a bit reluctant to do, as it would involve questionable heuristics that may not get me very far, in the long run.

Thank you.

**Comment by dvarrazzo** [comment]
Hello,

if there is no communication with the server the client will not get any information about the service closed. This is by design according to the PostgreSQL developer. You need to run a query or just to call `connection.poll()`. Recent psycopg versions set cnn.closed to 2 when the connection is closed by an error.

I've noticed that a single call to poll is currently failing to detect the broken connection. After a single call to `poll` there is only a notice raised but all the libpq functions run no problem. Even the state of the libpq connection seem regular:

    [server stopped here]

    In [10]: cnn.poll()
    [11680] conn_poll: status = 1
    [11680] conn_poll: async_status = ASYNC_DONE
    [11680] pq_is_busy: consuming input
    [11680] conn_notice_callback: FATAL:  terminating connection due to administrator command

    [11680] conn_notice_process: FATAL:  terminating connection due to administrator command

    [11680] conn_poll: poll reading
    [11680] conn_poll: async_status -> ASYNC_DONE
    Out[10]: 0

    In [11]: cnn.get_transaction_status()
    Out[11]: 0

a further call to `poll()` correctly detects the transaction as broken:

    In [13]: cnn.poll()
    [11680] conn_poll: status = 1
    [11680] conn_poll: async_status = ASYNC_DONE
    [11680] pq_is_busy: consuming input
    [11680] pq_is_busy: PQconsumeInput() failed
    [11680] conn_poll: poll reading
    ---------------------------------------------------------------------------
    OperationalError                          Traceback (most recent call last)
    <ipython-input-13-2d99764f7d0a> in <module>()
    ----> 1 cnn.poll()

    OperationalError: server closed the connection unexpectedly
    	This probably means the server terminated abnormally
    	before or while processing the request.


    In [14]: cnn.get_transaction_status()
    Out[14]: 4

    In [15]: cnn.closed
    Out[15]: 2

This happens with libpq 9.4.1. It is something to be investigated. However you seem to think that without interaction between client and server the client will eventually receive a disconnection message: no, this is not going to happen.

**Comment by dvarrazzo** [comment]
Fixed with #443.
